### ISSUE TITLE
bug report: certificate does not work in SMTP over SSL but works over StartTLS

### ISSUE THREAD

**Issue by MohammedNoureldin** [docker-mailserver/docker-mailserver#3809]
### 📝 Preliminary Checks

- [X] I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

### 👀 What Happened?

String with version 13.0.0, when using Proxy Protocol, the certificate of the SMTP server cannot be fetched over SSL (465), but is accessible over StartTLS (587).

Using the exact same configuration without a single change except using the version **12.1.0 solves the issue**.

I tried 13.0.0, 13.2.0 and 13.3.0, all have the same issue.

```
openssl s_client -connect mail.lixron.com:465
# In version 12.1.0 the previous command receives the certificate.
# In version 13.x.0 the previous command throws this error:
CONNECTED(00000003)
20F0869FFFFF0000:error:0A000126:SSL routines:ssl3_read_n:unexpected eof while reading:../ssl/record/rec_layer_s3.c:303:
no peer certificate available

openssl s_client -connect mail.lixron.com:587 -starttls smtp
# In version 12.1.0 and 13.x.x. the previous command receives the certificate.
```

Disabling Proxy Protocol solves the issue.

### 👟 Reproduction Steps

I am not going to post the whole configuration to keep the focus on the proxy protocol, as it seems to not be related to the configuration itself, but to the version and proxy protocol. Though if needed I can do post my configuration.

I am configuring proxy protocol as shown in the documentation in [Kubernetes page](https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/kubernetes/#proxy-port-to-service-via-proxy-protocol).

As mentioned, the exact same configuration with exact same ingress configuration for proxy protocol was used for testing, the only thing that changed was the version of DMS.

### 🐋 DMS Version

v13.0.0, 13.2.0, 13.3.0

### 💻 Operating System and Architecture

Ubuntu 22.04, ARM64

**Comment by polarathene** [comment]
> postfix-master.cf: |
>     smtp/inet/postscreen_upstream_proxy_protocol=haproxy
>     submission/inet/smtpd_upstream_proxy_protocol=haproxy
>     smtps/inet/smtpd_upstream_proxy_protocol=haproxy

> Is it related to this? #3546

Yes. See my comment: https://github.com/docker-mailserver/docker-mailserver/issues/3546#issuecomment-1902569889

When you upgrade to a new major release, please pay attention to the breaking changes section. That should in the majority of cases convey any changes introduced that can introduce breakage when you upgrade.

In this case you carry a custom `postfix-master.cf` with an override to the `smtps` service/transport.
- v13 of DMS renamed that to the `submissions` service name in `/etc/services` (from Debian) that maps to port 465.
- v13 likewise updated the related docs with that advice, and raised this concern for `smtps` in `master.cf` being changed as a breaking change in release notes.

---

> I tried disabling Proxy Protocol from all DMS services, so just setting the config as if there is no Proxy Protocol, and of course, removed proxy protocol from my ingress Nginx settings. Which worked.

> Could someone explain this also please?

Have you [seen the related docs page](https://docker-mailserver.github.io/docker-mailserver/v13.3/examples/tutorials/mailserver-behind-proxy/#information)?

This explains that while the connection may work fine, the intent of the proxy protocol is to preserve the original client IP (_test this from an external system, and perhaps an internal one_).

It may vary depending on environment (_OS / container runtime / network configuration / etc_), but similar to an IPv6 host to an IPv4 only container, it has been known to rewrite the client/remote IP to a local one. 

Another example is a separate Roundcube container, multiple users connect via the web interface, and the RC container connects directly to DMS within the internal private subnet (docker bridge), without the original client IP being forwarded all mail users appear to originate from the same RC container, and this allows for someone to fail auth enough times when Fail2Ban is active to trigger a ban on the IP, blocking all legitimate users.

**Comment by MohammedNoureldin** [comment]
> When you upgrade to a new major release, please pay attention to the breaking changes section. That should in the majority of cases convey any changes introduced that can introduce breakage when you upgrade.

You are right, I did not pay big attention on the changelog because I am not upgrading, but rather installing a fresh installation, and skimmed the documentation quickly without noticing that this particular config for postfix was changed.

> Have you [seen the related docs page](https://docker-mailserver.github.io/docker-mailserver/v13.3/examples/tutorials/mailserver-behind-proxy/#information)?

I am already familiar with Proxy Protocol and I am using it to preserve the IP of the clients, because of the existence of NGINX reverse proxy, which exposes the mail service to the outside.

Though I still do not understand why disabling Proxy Protocol on the port 465 causes no difference for me. I was still able to the the correct IP of the clients, which is probably because the clients use IMAP and not SMTP.

### UPDATE

I think I tested it the wrong way, I have had to test it by trying to send a mail from another server, which will then be using SMTP to deliver the mail to my server, which will appear as my reverse proxy IP when I disable the Proxy Protocol. Am I right here?

**Comment by MohammedNoureldin** [comment]
OK it works now after updating my settings. I will close the issue to keep the issue tracker clean.

**Comment by polarathene** [comment]
> Sending a mail from a server to my DMS will then be using SMTP to deliver the mail to my server, which will appear as my reverse proxy IP when I disable the Proxy Protocol. Am I right here?

Yes.

If you have another internal service between DMS and the remote client, the IP source is from the intermediary service instead which is misleading. Additional configuration is required to preserve the real client IP which is the purpose of the Proxy Protocol here.

---

> Though I still do not understand why disabling Proxy Protocol on the port 465 causes no difference for me. I was still able to the the correct IP of the clients.
> I think I know what I misunderstood here. It is probably because clients use IMAP and not SMTP to access the mails.

Port 465 is for SMTP submission over implicit TLS.

You would use this if sending mail from a mail client like Thunderbird, or the alternative submission (explicit TLS) port 587.

It is possible that other factors could be involved in your environment where the issue described does not occur. We are just aware of the scenario where it does happen, and that it is common enough to document awareness of.

---

> skimmed the documentation quickly without noticing that this particular config for postfix was changed.

We can't do much about skimming through breaking changes that result in breakage from misconfiguration on your end 😜

Try avoid that, or at least double check when you run into problems as an initial troubleshooting step. This is [suggested in our debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/#steps-for-debugging-dms):

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/fa451c21-f4e7-43bb-af8e-9e605b64d7b4)

You seem to have cheekily omitted this requirement, haha:

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/d76c676b-6bfc-406b-862a-0736d31a87ad)

The intention was it'd guide users to minimize opening issues which are resolved by the advice there, but I understand we can be impatient / frustrated at times that it can be easier to just get a bug report out of the way in the meantime. Hopefully it is somewhat accomplishing that (_hard to tell :(_)

**Comment by MohammedNoureldin** [comment]
> We can't do much about skimming through breaking changes that result in breakage from misconfiguration on your end 😜

fair enough, was my fault :D

> You seem to have cheekily omitted this requirement, haha:

Actually I read it already, but I just felt I needed to remove it as a disclaimer that I did not follow it 100% :D, that is because I am not using docker at all, but K8s, which has relatively different approach. That is why I removed this confirmation :D

> I do appreciate your bug reports though, you provide plenty of details and keep updating us with your findings as you narrow it down which is great! 😎

Thanks!
